Rubberduck.VBEEditor/ComManagement/TypeLibs.cs

```diff
@@ -0,0 +1,707 @@
+//   The type library is LIVE information, so consider it a snapshot at that very moment when you are dealing with it
+//   Make sure you call TypeLibsAccessor_VBE.Dispose() as soon as you have done what you need to do with it.
+//   Once control returns back to the VBE, you must assume that all the ITypeLib/ITypeInfo pointers are now invalid.
+//
+// CURRENT LIMITATIONS:
+// At the moment, enums and UDTs are not exposed through the type libraries
+// Constants names are not available
+
+// IMPLEMENTATION DETAILS:
+// There are two significant bugs in the VBE typeinfos implementations that we have to work around.
+// 1)  Some implementations of ITypeInfo provided by the VBE will crash with an AV if you call IProvideClassInfo::GetClassInfo on them.
+//      And guess what method the CLR calls on all COM interop objects when creating a RCW?  You guessed it.
+//      So, we use an aggregation object, plus ITypeInfo and ITypeLib wrappers to circumvent this VBE bug.
+//
+// 2)  The ITypeInfo for base classes of UserForms crash with an AV if you call ITypeInfo::GetDocumentation(MEMBERID_NIL) to get the type name
+//     We've got to remember that the VBE didn't ever intend for us to get hold of these objects, so there will be little bugs.
+//     This bug is also resolved in the provided wrappers.
+//
+// All the extended functionality is exposed through the wrappers.
+
+namespace Rubberduck.VBEditor.ComManagement.TypeLibs
+{
+    public class StructHelper
+    {
+        public static T ReadComObjectStructure<T>(object comObj)
+        {
+            // Reads a COM object as a structure to copy its internal fields
+            if (Marshal.IsComObject(comObj))
+            {
+                var referencesPtr = Marshal.GetIUnknownForObject(comObj);
+                var retVal = StructHelper.ReadStructure<T>(referencesPtr);
+                Marshal.Release(referencesPtr);
+                return retVal;
+            }
+            else
+            {
+                throw new ArgumentException("Expected a COM object");
+            }
+        }
+
+        public static T ReadStructure<T>(IntPtr memAddress)
+        {
+            if (memAddress == IntPtr.Zero) return default(T);
+            return (T)Marshal.PtrToStructure(memAddress, typeof(T));
+        }
+
+        public static T ReadStructureSafe<T>(IntPtr memAddress)
+        {
+            if (memAddress == IntPtr.Zero) return default(T);
+
+            // FIXME add memory address validation here, using VirtualQueryEx
+            return (T)Marshal.PtrToStructure(memAddress, typeof(T));
+        }
+    }
+
+    // AggregateSingleInterface is used to ensure that a wrapped COM object only responds to a specific interface
+    // In particular, we don't want them to respond to IProvideClassInfo, which is broken in the VBE for some ITypeInfo implementations 
+    public class AggregateInterfacesWrapper : ICustomQueryInterface, IDisposable
```

Comment:
yes, I couldn't reuse purely because this one works on pointers (due to the implicit IProvideClassInfo problem), and the other on an RCW.   But, we could almost definitely change the other one to use pointers at some point.